File: sdks/java/reactor-sdk/src/main/java/inc/reactor/sdk/ReactorException.java
package inc.reactor.sdk;

import org.jspecify.annotations.Nullable;

/**
 * A Reactor failure, in terms a caller can act on.
 *
 * <p>The same object is what a failed call throws and what an error event delivers. That is
 * deliberate: two vocabularies for one failure is how a 401 during connect ended up being
 * {@code UNAUTHORIZED} and not recoverable on one path, and {@code CONNECTION_FAILED} and
 * recoverable on the other.
 *
 * <p>Unchecked, because none of these is a condition a caller can be forced to handle usefully at
 * every call site. Catch the subclass you can do something about — {@link RateLimitedException},
 * {@link DisconnectedException} — and let the rest travel.
 *
 * <p>{@link #isRecoverable()} is derived from {@link #code()} and never stored per site, so two
 * SDKs cannot disagree about whether a timeout is worth retrying.
 */
public class ReactorException extends RuntimeException {

    private static final long serialVersionUID = 1L;

    /**
     * The code a failure carries when nothing classified it.
     *
     * <p>Not an {@link ErrorCode} constant: it is the base class's own code, the one left when
     * the platform named none, and it has no typed subclass to catch.
     */
    public static final String INTERNAL_ERROR = "INTERNAL_ERROR";

    private final String code;
    private final @Nullable Integer status;
    private final @Nullable String operation;
    private final @Nullable Long retryAfterMs;

    /**
     * @param code the stable, matchable code — one of {@link ErrorCode}, or one the platform sent
     * @param message the human-readable message
     * @param status the HTTP status, when the failure came from one
     * @param operation which call failed, e.g. {@code "connect"}
     * @param retryAfterMs a backoff hint, when the platform sent one
     */
    protected ReactorException(
            String code,
            String message,
            @Nullable Integer status,
            @Nullable String operation,
            @Nullable Long retryAfterMs) {
        super(message);
        this.code = code;
        this.status = status;
        this.operation = operation;
        this.retryAfterMs = retryAfterMs;
    }

    /**
     * The typed exception for a code.
     *
     * <p>A code this SDK knows becomes its own class, so a caller can catch the one thing they can
     * act on. A code it does not know becomes a plain {@code ReactorException} carrying that code
     * unchanged — the platform's set is open-ended, and an unrecognised code is a failure that
     * could not be classified rather than a payload that failed to parse.
     *
     * @param code the stable code
     * @param message the human-readable message
     * @param status the HTTP status, when the failure came from one
     * @param operation which call failed
     * @param retryAfterMs a backoff hint, when the platform sent one
     * @return the exception for that code
     */
    public static ReactorException of(
            String code,
            String message,
            @Nullable Integer status,
            @Nullable String operation,
            @Nullable Long retryAfterMs) {
        return ErrorCode.of(code)
                .map(known -> known.create(message, status, operation, retryAfterMs))
                .orElseGet(() -> new ReactorException(code, message, status, operation, retryAfterMs));
    }

    /**
     * The stable code for this failure.
     *
     * <p>Usually one of {@link ErrorCode}'s, but the platform's set is open-ended: a rejected
     * control request, command or recording may carry a code this SDK has never heard of. An
     * unrecognised code is an error that could not be classified — never a parse failure.
     *
     * @return the code, never empty
     */
    public String code() {
        return code;
    }

    /**
     * Whether the same call could succeed later.
     *
     * <p>True is about the moment — a timeout, a 5xx, a transport that dropped. False is about the
     * request, and repeating it unchanged fails the same way. A code this SDK does not recognise is
     * not recoverable, which is the safe direction: never promise a retry will help.
     *
     * @return whether retrying is worth anything
     */
    public boolean isRecoverable() {
        return ErrorCode.isRecoverable(code);
    }

    /**
     * @return the HTTP status behind this failure, or {@code null} when it did not come from one
     */
    public @Nullable Integer status() {
        return status;
    }

    /**
     * @return which call failed, or {@code null} when the failure names none
     */
    public @Nullable String operation() {
        return operation;
    }

    /**
     * @return how long to wait before retrying, or {@code null} when the platform sent no hint
     */
    public @Nullable Long retryAfterMs() {
        return retryAfterMs;
    }

    @Override
    public String toString() {
        StringBuilder out = new StringBuilder(getClass().getName())
                .append(": [")
                .append(code)
                .append("] ")
                .append(getMessage());
        if (operation != null) {
            out.append(" (operation: ").append(operation).append(')');
        }
        if (status != null) {
            out.append(" (HTTP ").append(status).append(')');
        }
        return out.toString();
    }
}

File: sdks/android/reactor-sdk-android/src/main/kotlin/inc/reactor/sdk/android/ReactorException.kt
package inc.reactor.sdk.android

/**
 * A failure, in terms a caller can act on.
 *
 * The same object an operation throws is the one an error event delivers — `Reactor.errors` and
 * a failed `suspend` call give you this, so there is one thing to catch and one thing to match
 * on. Matching on message text is what every SDK above this boundary was reduced to before the
 * codes existed.
 *
 * [recoverable] is **derived from [code]**, never passed in per call site. Two SDKs disagreeing
 * about whether a timeout is worth retrying is the failure that produced two names for one
 * condition — `connect()` reporting `CONNECTION_FAILED` and recoverable on one channel and
 * `UNAUTHORIZED` and not on the other.
 */
public open class ReactorException internal constructor(
    /** One of [ErrorCode]'s wire values, or a code the platform sent that this SDK does not know. */
    public val code: String,
    message: String,
    /** The HTTP status, when the failure came from one. */
    public val status: Int? = null,
    /** Which call failed — `"connect"`, `"send_command"`. */
    public val operation: String? = null,
    /** How long to wait before retrying, when the platform said. */
    public val retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : Exception(message, cause) {
    /**
     * Whether the same call could succeed later, unchanged.
     *
     * True is about the moment rather than the request: the connection went, the network did not
     * answer, the platform was busy. Everything else describes the request, and repeating it
     * unchanged fails the same way.
     */
    public val recoverable: Boolean
        get() = ErrorCode.isRecoverable(code)

    override fun toString(): String =
        buildString {
            append(this@ReactorException::class.simpleName)
            append('(')
            append(code)
            status?.let { append(", status=").append(it) }
            operation?.let { append(", operation=").append(it) }
            append("): ")
            append(message)
        }
}

/** The client is not in a state where this call means anything — publishing before connecting. */
public class InvalidStateException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** The session is gone. Recoverable: reconnect and try again. */
public class DisconnectedException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** The network did not answer. */
public class NetworkException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** The reply did not arrive in time. */
public class RequestTimeoutException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** The WebRTC transport failed. */
public class TransportException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** The key or token was rejected. Not recoverable: the same credentials fail the same way. */
public class UnauthorizedException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** No such model, session or clip. */
public class NotFoundException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** The request conflicts with the session's current state. */
public class ConflictException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** Too many requests. Recoverable, and [retryAfterMs] says when. */
public class RateLimitedException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** The request was malformed. */
public class BadRequestException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** The platform failed. Recoverable: a 5xx is about the moment. */
public class ServerException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** This SDK and the platform disagree about the protocol. */
public class VersionMismatchException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/**
 * A payload did not parse.
 *
 * Distinct from an absent payload, which is an answer. A successful completion whose result will
 * not decode is this, never an empty object: substituting `{}` makes a schema request answer with
 * a schema declaring nothing, which no caller can tell from a model that declares nothing.
 */
public class DecodeFailedException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** The session ended and will not come back. */
public class SessionTerminalException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** The message exceeded the channel's limit. */
public class MessageTooLargeException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** The operation was cancelled or the client went away before it finished. */
public class AbortedException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/** Recording is not enabled for this model or session. */
public class RecorderDisabledException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)

/**
 * A code this SDK does not know.
 *
 * The platform can reject a control request, command or recording with a code of its own, so the
 * set is open-ended. An unrecognised code is an error you cannot classify — never a parse
 * failure, and never silently reshaped into one of the known ones.
 */
public class UnknownReactorException internal constructor(
    code: String,
    message: String,
    status: Int? = null,
    operation: String? = null,
    retryAfterMs: Long? = null,
    cause: Throwable? = null,
) : ReactorException(code, message, status, operation, retryAfterMs, cause)
